src/object/object_primitive.c

```diff
@@ -16374,3 +16377,224 @@ mr_get_compression_length (const char *string, int charlen)
 
   return length;
 }
+
+
+/*
+ * or_write_string_to_buffer ()	  : Writes a VARCHAR or VARNCHAR to buffer, without needing the buffer initialised.
+ *				    
+ * buf(out)			  : Buffer to be written to.
+ * val_p(in)			  : Memory area to be written to.
+ * value(in)			  : DB_VALUE to be written.
+ * val_size(out)		  : Disk size of the DB_VALUE.
+ * align(in)			  : 
+ *
+ *  Note:
+ *	We use this to avoid double compression when it is required to have the size of the DB_VALUE, previous
+ *	to the write of the DB_VALUE in the buffer.
+ */
+
+int
+mr_write_string_to_buffer (OR_BUF * buf, char *val_p, DB_VALUE * value, int *val_size, int align)
+{
+  char *compressed_string = NULL, *string = NULL, *str = NULL;
+  int rc = NO_ERROR, str_length = 0, length = 0;
+  lzo_uint compression_length = 0;
+  lzo_bytep wrkmem = NULL;
+  bool compressed = false;
+  int error_abort = 0;
+
+  /* Checks to be sure that we have the correct input */
+  assert (DB_VALUE_DOMAIN_TYPE (value) == DB_TYPE_VARNCHAR || DB_VALUE_DOMAIN_TYPE (value) == DB_TYPE_STRING);
+  assert (DB_GET_STRING_SIZE (value) >= PRIM_MINIMUM_STRING_LENGTH_FOR_COMPRESSION);
+
+  error_abort = buf->error_abort;
+  buf->error_abort = 0;
+  string = DB_GET_STRING (value);
+  str_length = DB_GET_STRING_SIZE (value);
+  *val_size = 0;
+
+  /* Step 1 : Compress, if possible, the dbvalue */
+  wrkmem = (lzo_voidp) malloc (LZO1X_1_MEM_COMPRESS);
+  if (wrkmem == NULL)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, (size_t) LZO1X_1_MEM_COMPRESS);
+      goto cleanup;
+    }
+
+  memset (wrkmem, 0x00, LZO1X_1_MEM_COMPRESS);
+
+  /* Alloc memory for the compressed string */
+  /* Worst case LZO compression size from their FAQ */
+  compressed_string = malloc (str_length + (str_length / 16) + 64 + 3);
```

Comment:
What about using a macro for compressed string size ?
#define LZO_COMPRESSED_STRING_SIZE(str_length)   (str_length + (str_length / 16) + 64 + 3)